python/ray/air/preprocessor.py

```diff
@@ -158,25 +172,101 @@ def _check_is_fitted(self) -> bool:
+        for best performance.
+        * Implementation is defined as overriding the method in a sub-class.
+        """
+
+        assert data_format in ("pandas", "arrow")
+        has_transform_arrow = (
+            self.__class__._transform_arrow != Preprocessor._transform_arrow
+        )
+        has_transform_pandas = (
+            self.__class__._transform_pandas != Preprocessor._transform_pandas
+        )
+
+        if has_transform_arrow and has_transform_pandas:
+            # Both transforms available, so we delegate based on the dataset format to
+            # ensure minimal data conversions.
+            if data_format == "pandas":
+                transform_type = "pandas"
+            elif data_format == "arrow":
+                transform_type = "arrow"
+        elif has_transform_pandas:
+            transform_type = "pandas"
+        elif has_transform_arrow:
+            transform_type = "arrow"
+        else:
+            raise NotImplementedError(
+                "Neither `_transform_arrow` nor `_transform_pandas` are implemented."
+            )
+        return transform_type
+
     def _transform(self, dataset: Dataset) -> Dataset:
         # TODO(matt): Expose `batch_size` or similar configurability.
         # The default may be too small for some datasets and too large for others.
-        return dataset.map_batches(self._transform_pandas, batch_format="pandas")
+
+        dataset_format = dataset._dataset_format()
+        if dataset_format not in ("pandas", "arrow"):
+            raise ValueError(
+                f"Unsupported Dataset format: '{dataset_format}'. Only 'pandas' and "
+                "'arrow' Dataset formats are supported."
+            )
+
+        transform_type = self._determine_transform_to_use(dataset_format)
+
+        if transform_type == "pandas":
+            return dataset.map_batches(self._transform_pandas, batch_format="pandas")
+        else:
+            return dataset.map_batches(self._transform_arrow, batch_format="pyarrow")
 
     def _transform_batch(self, df: DataBatchType) -> DataBatchType:
         import pandas as pd
 
-        # TODO(matt): Add `_transform_arrow` to use based on input type.
-        # Reduce conversion cost if input is in Arrow:  Arrow -> Pandas -> Arrow.
-        if not isinstance(df, pd.DataFrame):
+        try:
+            import pyarrow
+        except ImportError:
+            pyarrow = None
+
+        if isinstance(df, pd.DataFrame):
+            data_format = "pandas"
+        elif pyarrow is not None and isinstance(df, pyarrow.Table):
+            data_format = "arrow"
+        else:
             raise NotImplementedError(
-                "`transform_batch` is currently only implemented for Pandas DataFrames."
+                "`transform_batch` is currently only implemented for Pandas DataFrames "
+                f"and PyArrow Tables. Got {type(df)}."
             )
```

Comment:
Double checking, should the Predictor already have converted numpy arrays to pandas by this point?

Reply:
Not yet - and if it did, it would fail in current CI :)